File: frontend/src/game/managers/SocketManager.ts
import type { Team, Position, PlayerStatus, FlagStatus } from '@/types'
import { WorldManager } from './WorldManager'
import { SocketConnectionManager } from './socket/SocketConnectionManager'
import { SocketMessageHandler } from './socket/SocketMessageHandler'
import { MessageSender } from './socket/MessageSender'
import { EventEmitter } from './socket/EventEmitter'

export interface GameInitParams {
  mapWidth: number
  mapHeight: number
  walls: Array<Position & { tileId?: number }>
  obstacles1: Position[]
  obstacles2: Position[]
  lteamPrison: Position[]
  lteamTarget: Position[]
  rteamPrison: Position[]
  rteamTarget: Position[]
}

export interface GameStatusParams {
  time: number
  lteamPlayerStatus: PlayerStatus[]
  lteamFlagStatus: FlagStatus[]
  rteamPlayerStatus: PlayerStatus[]
  rteamFlagStatus: FlagStatus[]
}

export enum SocketEvent {
  CONNECT = 'connect',
  DISCONNECT = 'disconnect',
  MESSAGE = 'message',
  ERROR = 'error',
  ACTIONS_RECEIVED = 'actions_received'
}

type EventListener = (...args: unknown[]) => void

export class SocketManager {
  private static instance: SocketManager | null = null
  private connectionManager: SocketConnectionManager
  private messageHandler: SocketMessageHandler
  private messageSender: MessageSender
  private emitter: EventEmitter

  private constructor(world: WorldManager) {
    this.emitter = new EventEmitter()
    this.connectionManager = new SocketConnectionManager(this.emitter)
    this.messageHandler = new SocketMessageHandler(this.emitter)
    this.messageSender = new MessageSender(world, this.connectionManager)

    this.emitter.on(SocketEvent.MESSAGE, (team: Team, data: unknown) => {
      this.messageHandler.handleMessage(team, data)
    })
  }

  static getInstance(world: WorldManager): SocketManager {
    if (!SocketManager.instance) {
      SocketManager.instance = new SocketManager(world)
    }
    return SocketManager.instance
  }

  /**
   * 获取连接管理器（用于测试）
   */
  getConnectionManager(): SocketConnectionManager {
    return this.connectionManager
  }

  connectTeam(team: Team, url: string): void {
    this.connectionManager.connectTeam(team, url)
  }

  disconnectTeam(team: Team): void {
    this.connectionManager.disconnectTeam(team)
  }

  sendGameInit(params: GameInitParams): void {
    this.messageSender.sendGameInit(params)
  }

  sendGameStatus(params: GameStatusParams): void {
    this.messageSender.sendGameStatus(params)
  }

  sendGameFinished(): void {
    this.messageSender.sendGameFinished()
  }

  isConnected(team: Team): boolean {
    return this.connectionManager.isConnected(team)
  }

  on(event: SocketEvent, listener: EventListener): void {
    this.emitter.on(event, listener)
  }

  off(event: SocketEvent, listener: EventListener): void {
    this.emitter.off(event, listener)
  }

  disconnectAll(): void {
    this.connectionManager.disconnectAll()
    this.emitter.removeAllListeners()
  }

  getConnectionStatus(): Record<Team, boolean> {
    return this.connectionManager.getConnectionStatus()
  }
}

File: native/managers/socket/socket_manager.py
"""SocketManager - singleton for WebSocket connections"""

from typing import Optional, Dict, Any
from .event_emitter import EventEmitter, SocketEvent, EventListener
from .team_socket import TeamSocket
from ...utils import Team


class SocketManager:
    """Socket manager (singleton)"""
    _instance: Optional['SocketManager'] = None

    def __new__(cls):
        if cls._instance is None:
            cls._instance = super().__new__(cls)
        return cls._instance

    def __init__(self):
        if hasattr(self, '_initialized'):
            return
        self.sockets: Dict[Team, TeamSocket] = {}
        self.emitter = EventEmitter()
        self._initialized = True

    def connect_team(self, team: Team, url: str):
        if team in self.sockets:
            self.disconnect_team(team)
        socket = TeamSocket(url, team, self.emitter)
        self.sockets[team] = socket
        socket.connect()

    def disconnect_team(self, team: Team):
        socket = self.sockets.get(team)
        if socket:
            socket.disconnect()
            self.sockets.pop(team, None)

    def send_game_init(self, params: Dict[str, Any]):
        map_payload = self._build_map_payload(params)
        if self.is_connected(Team.LEFT):
            self._send_to_team(Team.LEFT, self._build_init_payload("L", map_payload, params, True))
        if self.is_connected(Team.RIGHT):
            self._send_to_team(Team.RIGHT, self._build_init_payload("R", map_payload, params, False))

    def _build_map_payload(self, params: Dict[str, Any]) -> Dict[str, Any]:
        return {"width": params["map_width"], "height": params["map_height"],
                "walls": [{"x": w["x"], "y": w["y"]} for w in params["walls"]],
                "obstacles": [{"x": w["x"], "y": w["y"]} for w in params["obstacles1"]] +
                             [{"x": w["x"], "y": w["y"]} for w in params["obstacles2"]]}

    def _build_init_payload(self, team_name: str, map_payload: Dict, params: Dict, is_left: bool) -> Dict:
        my, opp = ("lteam", "rteam") if is_left else ("rteam", "lteam")
        return {"action": "init", "map": map_payload, "numPlayers": params["num_players"],
                "numFlags": params["num_flags"], "myteamName": team_name,
                "myteamPrison": params[f"{my}_prison"], "myteamTarget": params[f"{my}_target"],
                "opponentPrison": params[f"{opp}_prison"], "opponentTarget": params[f"{opp}_target"]}

    def send_game_status(self, params: Dict[str, Any]):
        if self.is_connected(Team.LEFT):
            self._send_to_team(Team.LEFT, self._build_status_payload(params, True))
        if self.is_connected(Team.RIGHT):
            self._send_to_team(Team.RIGHT, self._build_status_payload(params, False))

    def _build_status_payload(self, params: Dict[str, Any], is_left: bool) -> Dict:
        my, opp = ("lteam", "rteam") if is_left else ("rteam", "lteam")
        return {"action": "status", "time": params["time"], "myteamName": "L" if is_left else "R",
                "myteamPlayer": params[f"{my}_player_status"], "myteamFlag": params[f"{my}_flag_status"],
                "myteamScore": params[f"{my}_score"], "opponentPlayer": params[f"{opp}_player_status"],
                "opponentFlag": params[f"{opp}_flag_status"], "opponentScore": params[f"{opp}_score"]}

    def send_game_finished(self, lteam_score: int, rteam_score: int):
        if self.is_connected(Team.LEFT):
            self._send_to_team(Team.LEFT, {"action": "finished", "myteamScore": lteam_score, "opponentScore": rteam_score})
        if self.is_connected(Team.RIGHT):
            self._send_to_team(Team.RIGHT, {"action": "finished", "myteamScore": rteam_score, "opponentScore": lteam_score})

    def _send_to_team(self, team: Team, payload: Dict[str, Any]) -> bool:
        socket = self.sockets.get(team)
        return socket.send(payload) if socket else False

    def is_connected(self, team: Team) -> bool:
        socket = self.sockets.get(team)
        return socket.is_connected() if socket else False

    def on(self, event: SocketEvent, listener: EventListener):
        self.emitter.on(event, listener)

    def off(self, event: SocketEvent, listener: EventListener):
        self.emitter.off(event, listener)

    def emit(self, event: SocketEvent, *args, **kwargs):
        self.emitter.emit(event, *args, **kwargs)

    def disconnect_all(self):
        for socket in list(self.sockets.values()):
            socket.disconnect()
        self.sockets.clear()
        self.emitter.remove_all_listeners()

    def get_connection_status(self) -> Dict[Team, bool]:
        return {Team.LEFT: self.is_connected(Team.LEFT), Team.RIGHT: self.is_connected(Team.RIGHT)}
